### DISCUSSION TITLE
How to nest listing numbers within a sections/subsections?

### DISCUSSION THREAD

**Discussion by mlamias** [quarto-dev/quarto-cli#5402] [Q&A]
NOTE:  This is a cross post from https://stackoverflow.com/questions/76165783/nested-figure-numbers-in-quarto-html-document.

I'm creating an HTML Quarto document, where I have chapter numbers (e.g. Chapter 1), and then sections (e.g. 1.1, 1.2, 1.3, etc.), and sometimes sub-sections numbered, for example, 1.1.1, 1.1.2, 2.1.1. 2.1.2, 2.4.3, etc.  Within the sections or subsections, I frequently display figure/listing output and am automatically applying caption numbering, but the caption numbers seem to repeat the section number or subsection number (e.g. Figure 1.1 for figures in Section 1.1).  How can I start figure/listing caption numbers as X.Y1.Y2.Z, where X is the chapter number, Y1 is the section number, Y2 is the (optional) subsection number, and Z is sequentially numbered and restarts in each section or subsection.  

When I currently apply listing/figure numbering, this creates the following output:

> Chapter 1
> 
> 1.1 My First Section
> 
> Figure 1.1 <Figure here>
> 
> Chapter 2
> 
> 2.1  My First Section in Chapter 2
> 
> 2.2  My Second Section in Chapter 2
> 
> 2.2.1 My First Subsection in Section 2, Chapter 2
> 
> Figure 2.2 <Figure here>
> 
> Figure 2.3 <Figure here>

Instead, what I'd like is the following:

> Chapter 1
> 
> 1.1 My First Section
> 
> Figure 1.1.1 <Figure here>
> 
> Chapter 2
> 
> 2.1  My First Section in Chapter 2
> 
> 2.2  My Second Section in Chapter 2
> 
> 2.2.1 My First Subsection in Section 2, Chapter 2
> 
> Figure 2.2.1.1 <Figure here>
> 
> Figure 2.2.1.2 <Figure here>

Does anyone have any idea how I can change the default numbering to nest the figure number into the section or subsection number?  If this can't be accomplished directly in Quarto, is there a way to accomplish this with CSS, perhaps?

Thanks.

**Comment by mcanouil** [comment]
CSS won't help.
You could possibly write a LUA filter to parse the document and change numbering of cross-referenceable elements.

**Comment by cderv** [comment]
It seems like some improvement idea for our crossref system. 

@cscheid do you think numbering for crossref system could be improve to take into account more sub section ? Maybe something for 1.5 ?

**Reply by mlamias** [reply]
I'd love to see this as an option in 1.5.  It seems really confusing to me to have a Figure or Listing called something like 2.3 in a section labeled 2.1.4.  It seems that as of now, the numbering starts sequentially at the beginning of each chapter and only extends to a single hierarchical number level (e.g. 2.X).  Would love to see this improved.  

Thanks, @mcanouil for the suggestion too.  I toyed with the idea of Lua filters, but I'd need to brush up on the Lua syntax and would have to invest some time in learning that a bit more.  I'll take a look at it to see how difficult it might be to get up to speed on Lua filters, which certainly aren't in my current expertise.

**Reply by mlamias** [reply]
I just wanted to follow-up to see if this has possibly been added as an improvement for 1.5?  Any luck @cscheid?

**Reply by cscheid** [reply]
> I'd love to see this as an option in 1.5. 

(To clarify: we're working towards 1.4 right now) 

> It seems really confusing to me to have a Figure or Listing called something like 2.3 in a section labeled 2.1.4. 

I believe your experience, but this is how most cross-referencing systems work in existing books (I just checked my bookshelf) and in the crossreferencing system for LaTeX as well. Quarto is simply using current practices.

> Would love to see this improved.

There are technical reasons for why this isn't as simple as it seems, so it's not a sure shot. And, FYI, when we have improvements, we announce them on the [changelog](https://github.com/quarto-dev/quarto-cli/blob/main/news/changelog-1.4.md). Pinging us ends up slowing down the overall process, because we try to respond to every message, and there are a lot of you! Please consider not doing that. If you want to keep track where we're working on crossrefs, see #4944.

**Reply by mlamias** [reply]
Got it.  Thanks.  Sorry about the ping.  I wasn't aware of where the upcoming changelog kept or where you're tracking crossrefs.  Really appreciate that and will keep an eye on that in the future!  Thanks again!
